components/inventory.ts

```diff
@@ -76,15 +99,95 @@ export function createInventory(
         return inventoryUserCache.get(profileId)!
     }
 
+    // Get user data to check on location progression
+    const userData = getUserData(profileId, gameVersion)
+
     // add all unlockables to player's inventory
     const allunlockables = getVersionedConfig<Unlockable[]>(
         "allunlockables",
         gameVersion,
         true,
     ).filter((u) => u.Type !== "location") // locations not in inventory
 
+    let unlockables: Unlockable[] = allunlockables
+
+    if (getFlag("enableMasteryProgression")) {
+        // list of all items that are unlockable via package, like starter items
+        const packageUnlockables = allunlockables.reduce((acc, unlockable) => {
+            if (unlockable.Type === "package") {
+                acc.push(...unlockable.Properties.Unlocks)
+            }
+            return acc
+        }, [])
+
+        const challengesUnlockables =
+            controller.challengeService.getChallengesUnlockables()
+
+        /**
+         * Seperates unlockable types and looksup for progression level
+         * on unlockables that are locked behind mastery progression level
+         */
+        const [unlockedItems, otherItems]: [Unlockable[], Unlockable[]] =
```

Comment:
I have a feeling this in combination with getMasteryForUnlockable can be done in a more performant way. /createInventory is called quite often when going through the menu, and constantly looping through ALL unlockable and looping through each mastery everytime is a big-O of n*n at worst (for each unlockable, loop through each mastery until we find what we need).

Reply:
Ive removed the clearInventory from the end mission, so this will only ever be once when the user logs in and fetched from cache every other time, so it should avoid any big-O operations. The drop awards are done by injecting the drops directly into the cache whenever its required

Reply:
I do get the point for the getMasteryForUnlockable tho, I can probably map the data outside the functions, so its more easily accessable via objectIndex, so I dont have to loop trough everything again. Ill have a look